In [1]:
from os import chdir 
chdir('./lib')

In [2]:
!pwd

/Users/ginodefalco/dsi/dsi_repo/DSI_SM_3/projects/project-05/lib


In [3]:
from project_5 import general_process, load_data_from_database, make_data_dict, general_model, general_transformer

# Step 2 - Identify Salient Features Using $\ell1$-penalty

**NOTE: EACH OF THESE SHOULD BE WRITTEN SOLELY WITH REGARD TO STEP 2 - Identify Features**

### Domain and Data

MADELON is an artificial dataset, which was part of the NIPS 2003 feature selection challenge. This is a two-class classification problem with continuous input variables. The difficulty is that the problem is multivariate and highly non-linear.
MADELON is an artificial dataset containing data points grouped in 32 clusters placed on the vertices of a five dimensional hypercube and randomly labeled +1 or -1. The five dimensions constitute 5 informative features. 15 linear combinations of those features were added to form a set of 20 (redundant) informative features. Based on those 20 features one must separate the examples into the 2 classes (corresponding to the +-1 labels). We added a number of distractor feature called 'probes' having no predictive power. The order of the features and patterns were randomized.

### Problem Statement

Now that we have a benchmark established, we can can make some tweaks in order to filter our features.

### Solution Statement

We will do this by using the LASSO parameter within the Logistic Regression

### Metric

Our metric will be the same...our accuracy score from our train and test data.

### Benchmark

Our benchmark was a 78.8% training score and 51.6% test score.

## Implementation

Implement the following code pipeline using the functions you write in `lib/project_5.py`.

<img src="assets/identify_features.png" width="600px">

In [4]:
# Let's use the function we wrote to load our data into the dataframe
df = load_data_from_database()
df.shape

(2000, 502)

In [5]:
# Let's get info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Columns: 502 entries, index to label
dtypes: int64(502)
memory usage: 7.7 MB


In [6]:
# Describe the Data
df.describe()

index     feat_000     feat_001     feat_002     feat_003  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    999.500000   481.722500   483.452500   510.166000   483.384500   
std     577.494589     6.421769    30.186294    38.899165     9.059895   
min       0.000000   462.000000   381.000000   370.000000   453.000000   
25%     499.750000   477.000000   464.000000   485.000000   477.000000   
50%     999.500000   482.000000   483.000000   510.500000   483.000000   
75%    1499.250000   486.000000   503.000000   536.000000   490.000000   
max    1999.000000   503.000000   600.000000   654.000000   519.000000   

          feat_004     feat_005     feat_006     feat_007     feat_008  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    501.612500   479.259000   480.109500   476.565000   486.793500   
std      41.389418     6.795956    40.575925     1.384461    15.043836   
min     371.000000   459.000000   334.000000   471.000000   430.000000   
25%     475.000000   475.000000   452.750000   476.000000   477.000000   
50%     500.000000   479.000000   480.000000   477.000000   487.000000   
75%     528.000000   484.000000   506.250000   477.000000   496.250000   
max     688.000000   505.000000   611.000000   481.000000   536.000000   

          ...         feat_491     feat_492     feat_493    feat_494  \
count     ...      2000.000000  2000.000000  2000.000000  2000.00000   
mean      ...       478.811500   486.356500   496.565500   493.49950   
std       ...         4.011735    23.967366   127.635442    34.81902   
min       ...       463.000000   391.000000   130.000000   368.00000   
25%       ...       476.000000   471.000000   404.000000   470.00000   
50%       ...       479.000000   486.000000   504.000000   492.00000   
75%       ...       481.000000   502.000000   586.000000   517.00000   
max       ...       497.000000   566.000000   920.000000   615.00000   

          feat_495     feat_496     feat_497     feat_498     feat_499  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    510.893000   478.219500   483.309000   507.977000   490.266000   
std      37.459353     5.880613    13.559847    37.224297    25.825273   
min     398.000000   457.000000   435.000000   363.000000   403.000000   
25%     486.000000   474.000000   474.000000   482.000000   473.000000   
50%     511.000000   478.000000   483.000000   508.000000   490.000000   
75%     535.000000   482.000000   492.000000   533.000000   507.250000   
max     661.000000   500.000000   535.000000   644.000000   583.000000   

            label  
count  2000.00000  
mean      0.00000  
std       1.00025  
min      -1.00000  
25%      -1.00000  
50%       0.00000  
75%       1.00000  
max       1.00000  

[8 rows x 502 columns]

In [7]:
# Let's load our make_data_dict function to 
df = make_data_dict(df, random_state=0)
df

{'X_test':       index  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
 533     533       489       510       472       475       454       485   
 1369   1369       480       509       518       489       434       483   
 1802   1802       485       430       435       488       420       493   
 600     600       479       516       501       487       495       480   
 588     588       488       482       520       485       452       473   
 843     843       472       470       480       484       533       472   
 1377   1377       482       550       523       483       518       479   
 1828   1828       484       487       565       477       509       486   
 60       60       486       441       492       490       528       474   
 801     801       491       464       527       489       491       487   
 589     589       482       491       515       484       524       479   
 779     779       471       474       500       478       491       500   
 1

In [8]:
# Now to scale our data
from sklearn.preprocessing import StandardScaler
df = general_transformer(StandardScaler(), df)

In [9]:
df

{'X_test': array([[-0.80543598,  1.13394378,  0.90418143, ...,  0.13812851,
         -0.63386416,  1.00165037],
        [ 0.64776067, -0.26932334,  0.87060785, ...,  1.02293745,
         -0.36710877, -1.09829994],
        [ 1.40043309,  0.5102695 , -1.78170544, ...,  0.06439443,
          0.05969986,  0.37944287],
        ..., 
        [ 0.78682255, -0.42524191,  0.87060785, ...,  1.90774639,
          0.11305094,  1.62385788],
        [-1.61373316, -0.73707905,  0.70273992, ..., -0.15680781,
          0.75326388,  0.06833912],
        [ 0.64254585,  1.28986235,  2.11283053, ..., -0.45174412,
          0.0863754 ,  0.37944287]]),
 'X_train': array([[ 1.1570748 , -0.89299762,  0.43415123, ...,  0.94920337,
          0.75326388, -0.24276463],
        [-0.47342574,  0.19843237,  0.80346068, ...,  1.31787377,
          0.19307755, -0.67053228],
        [-1.02793499, -0.58116048, -1.84885261, ..., -0.59921228,
          0.59321064, -0.98163603],
        ..., 
        [ 0.70338542, -0.425241

In [10]:
# We are running a simple logistic regression with the LASSO to identify salient features
from sklearn.linear_model import LogisticRegression
df = general_model(LogisticRegression(C=1, penalty='l1'), df)

In [11]:
df

{'X_test': array([[-0.80543598,  1.13394378,  0.90418143, ...,  0.13812851,
         -0.63386416,  1.00165037],
        [ 0.64776067, -0.26932334,  0.87060785, ...,  1.02293745,
         -0.36710877, -1.09829994],
        [ 1.40043309,  0.5102695 , -1.78170544, ...,  0.06439443,
          0.05969986,  0.37944287],
        ..., 
        [ 0.78682255, -0.42524191,  0.87060785, ...,  1.90774639,
          0.11305094,  1.62385788],
        [-1.61373316, -0.73707905,  0.70273992, ..., -0.15680781,
          0.75326388,  0.06833912],
        [ 0.64254585,  1.28986235,  2.11283053, ..., -0.45174412,
          0.0863754 ,  0.37944287]]),
 'X_train': array([[ 1.1570748 , -0.89299762,  0.43415123, ...,  0.94920337,
          0.75326388, -0.24276463],
        [-0.47342574,  0.19843237,  0.80346068, ...,  1.31787377,
          0.19307755, -0.67053228],
        [-1.02793499, -0.58116048, -1.84885261, ..., -0.59921228,
          0.59321064, -0.98163603],
        ..., 
        [ 0.70338542, -0.425241

# Step 2 Conclusion:

We built a pipeline with LogisticRegression using l1(LASSO) penalty
and used our constructed model to identify important features. Our score this time was:
79% Train
54% Test
 - A slight improvement in identifying our features
 